### PR TITLE
Replace "run now" CustomAction with standard action PerformSingleExecution

### PR DESCRIPTION
**Pull request by mgoworko**
## Describe your changes

The main purpose of this PR, is to refactor "run now" action for periodic DeploymentManagers:
- before: it was a CustomAction
- now: it is a standard actions, similar to CANCEL, DEPLOY, etc.

Planned changes related to the PeriodicDeploymentManager:
1. This change modifies basic handling and UX of "run now" action
1A. Completely remove CustomAction mechanism, as it will be unused after stage 1 "run now"
3. Second phase (already in progress while writing this comment) will involve:
- moving the periodic deployment manager closer to Nu core
  -  database, that was separate, moved the main Nu db
  -  deployment manager decoupled from Flink
  -  interface moved to the deployment manager API module, but mostly as-is, without huge changes
- improving and optimizing queries about periodic scenarios and their status
4. Third phase may involve further integrating the periodics into the general deployment manager interface - phase 2 is preparation for that.
-----------------------------------------------------------------

Most important changes:

1. The "run now" actions is now implemented as standard action:

- it has its own dedicated button on `actions` panel, no longer uses generic CustomAction mechanism and button
- this new button is now fully controlled by Nu backend response, both
  - its visibility (only for periodic processes) 
  - active/inactive (greyed out), it is active only when newest version of scenario is scheduled

5. The ProcessStateDefinitionManager improvements:
- it is now aware, which version of the scenario is newest, and which is deployed
- it allows to manage allowed actions, depending on deployed and latest version (it was needed for "run now" action)
- ProcessStateDefinitionManager now allows to define, which action buttons are visible on FE (`def applicableActions`) (right now only applied to "run now" for periodic processes)
- ProcessStateDefinitionManager now allows to specify custom tooltips for action buttons (again, it was needed by "run now" action")

6. The PeriodicProcessStateDefinitionManager changes:
- "run now" is now a standard action, handled in the graph of state transitions
- "run now" is available only when newest version of scenario is scheduled
- in other states, custom tooltips are returned, describing why it is now possible to invoke "run now"

## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

## Summary by CodeRabbit

- **New Features**
	- Introduced `PerformSingleExecution` action, now supported in the DeploymentManager and GUI.
	- Added new command type `DMPerformSingleExecutionCommand` for processing single execution requests.
	- Enhanced process state management with new fields for visible actions and action tooltips.
	- New `PerformSingleExecutionButton` added to the toolbar for executing single actions.
	- Improved tooltip management for actions based on current process state.
	- Updated API endpoints for better error handling and specificity in responses.

- **Bug Fixes**
	- Improved error handling and logging for health check endpoints.

- **Documentation**
	- Updated changelog and migration guide to reflect new features and changes in command structure.
	- Added new test cases for tooltip visibility based on process states.

### CODE REVIEW

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/ProcessStateDefinitionManager.scala [hunk 1]
```diff
@@ -60,3 +79,22 @@ trait ProcessStateDefinitionManager {
   }
 
 }
+
+object ProcessStateDefinitionManager {
```
Comment (gskrobisz) [inline designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/ProcessStateDefinitionManager.scala:83]: Please add some documentation.

File: extensions-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/ProcessAction.scala [hunk 2]
```diff
@@ -71,6 +71,8 @@ object ScenarioActionName {
   val Pause: ScenarioActionName     = ScenarioActionName("PAUSE") // TODO: To implement in future..
   val Rename: ScenarioActionName    = ScenarioActionName("RENAME")
 
+  val PerformSingleExecution: ScenarioActionName = ScenarioActionName("run now")
```
Comment (gskrobisz) [inline extensions-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/ProcessAction.scala:74]: Please add comment and/or TODO why we keep "run now".

**Review by gskrobisz — APPROVED** [review]

File: extensions-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/ProcessAction.scala (hunk 2 above)
Comment (mgoworko) [reply extensions-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/ProcessAction.scala]: Comment added

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/ProcessStateDefinitionManager.scala (hunk 1 above)
Comment (mgoworko) [reply designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/ProcessStateDefinitionManager.scala:83]: Added docs for class and val below

File: designer/client/src/components/toolbarSettings/buttons/ActionButton.tsx [hunk 3]
```diff
@@ -17,20 +17,7 @@ export function ActionButton({ name, type }: ActionButtonProps): JSX.Element {
...
-
     return action ? (
-        <CustomActionButton
-            action={action}
-            processName={processName}
-            processStatus={status}
-            disabled={name === "run now" ? disabledValue : false}
-            type={type}
-        />
+        <CustomActionButton action={action} processName={processName} disabled={false} processStatus={status} type={type} />
```
Comment (arkadius) [inline designer/client/src/components/toolbarSettings/buttons/ActionButton.tsx:21]: I see that you've removed `CustomActionButton`. Why it is still used?

File: designer/client/src/components/toolbars/scenarioActions/buttons/PerformSingleExecutionButton.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    const isPossible = useSelector(isPerformSingleExecutionPossible);
+    const processName = useSelector(getProcessName);
+    const capabilities = useSelector(getCapabilities);
+    const available = !disabled && isPossible && capabilities.deploy;
+
+    const { open } = useWindows();
+    const action = (p, c) => HttpService.performSingleExecution(p, c).finally(() => dispatch(loadProcessState(processName)));
+    const message = t("panels.actions.perform-single-execution.dialog", "Perform single execution", { name: processName });
+
+    const tooltip = ProcessStateUtils.getActionCustomTooltip(scenarioState, PredefinedActionName.PerformSingleExecution) ?? "run now";
```
Comment (arkadius) [inline designer/client/src/components/toolbars/scenarioActions/buttons/PerformSingleExecutionButton.tsx:34]: Please use i18n mechanism

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DMScenarioCommand.scala [hunk 5]
```diff
@@ -86,3 +80,9 @@ case class DMCancelScenarioCommand(scenarioName: ProcessName, user: User) extend
 
 case class DMStopScenarioCommand(scenarioName: ProcessName, savepointDir: Option[String], user: User)
     extends DMScenarioCommand[SavepointResult]
+
+case class DMPerformSingleExecutionCommand(
```
Comment (arkadius) [inline designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DMScenarioCommand.scala:84]: 1. What about custom actions? Can you write in the description the plan of changes that will be made in further PRs? 
2. Let's write a TODO describing further plan that won't be a part of currently planned changes. Do you agree that eventually we should extract some kind of a scheduler interface and in the deployment manager keep only logic related with the deployment? Or maybe you see it differently?

**Review by arkadius — COMMENTED** [review]
Minor comments added

File: designer/client/src/reducers/graph/utils.fixtures.ts [hunk 6]
```diff
@@ -167,7 +167,9 @@ export const state: GraphState = {
                 name: "NOT_DEPLOYED",
             },
             version: null,
+            applicableActions: ["DEPLOY", "ARCHIVE", "RENAME"],
```
Comment (arkadius) [inline designer/client/src/reducers/graph/utils.fixtures.ts:170]: It is not easy to understand what is the difference between applicable vs allowed actions. Maybe we should refactor the original naming? How they different? One status make action buttons visible and another make them clickable (enabled)?

File: designer/client/src/components/toolbars/scenarioActions/buttons/PerformSingleExecutionButton.tsx (hunk 4 above)
Comment (mgoworko) [reply designer/client/src/components/toolbars/scenarioActions/buttons/PerformSingleExecutionButton.tsx]: I think I've done it now. Backend returns type of tooltip ("NOT_ALLOWED_IN_CURRENT_STATE"  /"NOT_ALLOWED_FOR_DEPLOYED_VERSION") and frontend handles actual message rendering with i18s.

File: designer/client/src/reducers/graph/utils.fixtures.ts (hunk 6 above)
Comment (mgoworko) [reply designer/client/src/reducers/graph/utils.fixtures.ts]: I renamed applicableActions to `visibleActions` and left the `allowedActions` unchanged. That way we have lists of `visible` and `allowed` actions, IMO that is clearer.

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DMScenarioCommand.scala (hunk 5 above)
Comment (mgoworko) [reply designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DMScenarioCommand.scala:84]: I added in this PR description the general plan for periodics, especially the change that is already in progress. 

I'm not sure ATM about the future of CustomAction's. We aren't using them internally anymore - "run now" was our only usage. They are left in the codebased and APIs right now, so Nu users can still use them. 

I guess we need to decide, whether to leave CustomAction's as they are, or remove them.

File: designer/client/src/components/toolbarSettings/buttons/ActionButton.tsx (hunk 3 above)
Comment (mgoworko) [reply designer/client/src/components/toolbarSettings/buttons/ActionButton.tsx:21]: The CustomActionButton was duplicated in the codebase, I removed the unused copy of that button. It is still available and used, when custom actions are defined for some DeploymentManager (we do not use this mechanism internally in Nu, "run now" was our only usage)

File: designer/client/src/components/toolbars/scenarioActions/buttons/PerformSingleExecutionButton.tsx (hunk 4 above)
Comment (arkadius) [reply designer/client/src/components/toolbars/scenarioActions/buttons/PerformSingleExecutionButton.tsx]: It's not exactly what I head in mind. I mean  i18n for "run now" text :). I would not apply i18n for backend messages this way. If we decide to do this, it should be done full on BE side and done more holistic. Let's revert this

Comment (mgoworko) [reply designer/client/src/components/toolbars/scenarioActions/buttons/PerformSingleExecutionButton.tsx]: Done, reverted changes related with backed-generated messages and added transaltion for "run now"

File: designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DMScenarioCommand.scala (hunk 5 above)
Comment (mgoworko) [reply designer/deployment-manager-api/src/main/scala/pl/touk/nussknacker/engine/api/deployment/DMScenarioCommand.scala:84]: Custom action removal added to the plan.

**Review by arkadius — APPROVED** [review]
`PeriodicProcessServiceIntegrationTest` failed. Is it related with your change?

**Comment by mgoworko** [comment]
> `PeriodicProcessServiceIntegrationTest` failed. Is it related with your change?

Seems completely not related. My change did not modify how periodic processes work internally. Logs are not conclusive, but I think it was Flink deployment error that caused it. It passes locally and passed on retry.

**Comment by arkadius** [comment]
> > `PeriodicProcessServiceIntegrationTest` failed. Is it related with your change?
> 
> Seems completely not related. My change did not modify how periodic processes work internally. Logs are not conclusive, but I think it was Flink deployment error that caused it. It passes locally and passed on retry.

OK, so let's merge it

**Pull request merged**